[ADD] estate: created a new module

hapat-odoo opened the pull request:
Created and configured the database for the Real Estate module, defining necessary fields such as property type, tags, buyer, salesperson, and offers. Implemented list, form, and search views with groupings, and added computed fields, inverse functions, and onchange functionalities. Included 'Sold' and 'Cancel' buttons with their corresponding functionality, as well as 'Accept' and 'Refuse' buttons with related logic.

hapat-odoo committed:
[ADD]  estate:  Added a New Module estate_property

Added a new module for estate property and completed tasks 1, 2, and 3.
 I added a models folder and learned to push data into PostgreSQL using ORM.

hapat-odoo committed:
[IMP] estate:  Implemented the Security Logic

Enhanced the Security Logic for improved protection. Added the access control
rights for the estate_property.py file

robodoo commented:
[![Pull request status dashboard](https://mergebot.odoo.com/odoo/tutorials/pull/1021.png)](https://mergebot.odoo.com/odoo/tutorials/pull/1021)

hapat-odoo committed:
[IMP] estate: Added menu structure and property action

Introduced a new menu structure for the estate module, including a root menu and submenus for properties. Also added an action for managing estate properties with a corresponding view mode.

hapat-odoo committed:
[IMP] estate:  menu structure and added property views

hapat-odoo committed:
[IMP] estate:  Added state field and search view

-added  search view state field and list and form views

[IMP] estate:  Added list & form view

-added  filter for new or offer received stage

[IMP] estate:  Implemented models and views for property offers

-added  models and views for property offers field

bit-odoo commented on Estate/models/estate_property.py:
```diff
@@ -0,0 +1,40 @@
+
```
unnecessary diff.

bit-odoo commented on Estate/models/estate_property.py:
```diff
@@ -0,0 +1,40 @@
+        copy=False,
+        default='new'
+    )
+    
```
An empty line should be without whitespace.

bit-odoo commented on Estate/views/estate_menus.xml:
```diff
@@ -0,0 +1,21 @@
+              name="List"
+              parent="estate_property_menu"
+              action="action_estate_property_list_only"/>
+
```
unnecessary diff.

bit-odoo commented on Estate/views/estate_menus.xml:
```diff
@@ -0,0 +1,21 @@
+    <menuitem id="estate_property_list_menu"
+              name="List"
+              parent="estate_property_menu"
+              action="action_estate_property_list_only"/>
```
```suggestion
    <menuitem id="estate_property_list_menu"
            name="List"
            parent="estate_property_menu"
            action="action_estate_property_list_only"/>
```

bit-odoo commented on Estate/views/estate_property_views.xml:
```diff
@@ -0,0 +1,88 @@
+        <field name="view_id" ref="estate_property_form_view"/>
+    </record>
+    
+
```
unnecessary diff.

bit-odoo commented on Estate/views/estate_property_views.xml:
```diff
@@ -0,0 +1,88 @@
+                <separator/>
+                <filter name="group_by_postcode"
+                        string="Group by Postcode"
+                        context="{'group_by': 'postcode'}"/>
```
Please check indentation.

bit-odoo commented on Estate/views/estate_menus.xml:
```diff
@@ -0,0 +1,21 @@
+
+    <menuitem id="estate_root_menu" name="Estate"/>
+
+    <menuitem id="estate_advertisement_menu"
+              name="Advertisements"
+              parent="estate_root_menu"/>
+    <menuitem id="estate_property_menu"
+              name="Properties"
+              parent="estate_advertisement_menu"
+              action="action_estate_property"/>
```
Same as below.

bit-odoo reviewed (commented):
Hello, Good Work!
I have left some comments.

For commit message you can refer this documentation :- https://www.odoo.com/documentation/19.0/contributing/development/git_guidelines.html#commit-message-full-description

hapat-odoo committed:
[IMP] estate: Implemented models and views for property offers, tags, and types

Implemented models and views for property offers
tags and types

hapat-odoo committed:
[IMP] estate: added automation logic

Enhanced property model and views with new fields and improved form layout

hapat-odoo committed:
[IMP] estate: Implemented offer acceptance and refusal actions

 Implemented offer acceptance and refusal actions;
added state management buttons in property views

[FIX] estate: resolve Runbot style and linting errors (Ruff/Semgrep)

This commit fixes all Runbot Check Style issues for the estate module.

[FIX] estate: Cleaned up code by removing unnecessary blank lines in model files

bit-odoo commented on Estate/models/estate_property.py:
```diff
@@ -0,0 +1,108 @@
+from odoo import models, fields, api
+from datetime import timedelta
```
```suggestion
from datetime import timedelta

from odoo import models, fields, api

```

bit-odoo commented on Estate/models/estate_property.py:
```diff
@@ -0,0 +1,108 @@
+        ("west", "West"),
+    ])
+    active = fields.Boolean(default=True)
+
```
unnecessary diff.

bit-odoo commented on Estate/models/estate_property.py:
```diff
@@ -0,0 +1,108 @@
+        ("sold", "Sold"),
+        ("canceled", "Canceled"),
+    ], string="Status", required=True, copy=False, default="new")
+
```
unnecessary diff.

bit-odoo commented on Estate/models/estate_property.py:
```diff
@@ -0,0 +1,108 @@
+    salesperson_id = fields.Many2one("res.users", string="Salesperson")
+    tag_ids = fields.Many2many("estate.property.tag", string="Tags")
+    offer_ids = fields.One2many("estate.property.offer", "property_id", string="Offers")
+
```
Same here.

hapat-odoo committed:
[IMP] estate:  Implemented  constraints for price validation

-Refactored validity field to validity_days
-added constraints for price validation

bit-odoo commented on Estate/models/estate_property.py:
```diff
@@ -0,0 +1,127 @@
+    _check_selling_price = models.Constraint(
+        'CHECK(selling_price < 0)',
+        'The selling price of a property must be positive.'
+    )
```
Constraints should be defined after the field declaration. And is this your condition going to work properly ?. Have you checked it?

bit-odoo commented on Estate/views/estate_property_views.xml:
```diff
@@ -0,0 +1,130 @@
+                    <button name="action_set_sold" string="Mark as Sold" type="object"/>
+                    <button name="action_set_canceled" string="Cancel" type="object"/>
+                    <field name="state" widget="statusbar" statusbar_visible="new,offer_received,offer_accepted,sold,canceled"/>
+                </header>
```
should be outside of the sheet for better visibility. You can take reference from other Odoo apps.

bit-odoo commented on Estate/models/estate_property.py:
```diff
@@ -0,0 +1,127 @@
+            if  record.selling_price < 0.9 * record.expected_price:
+                raise ValidationError("Selling price must be at least 90% of the expected price")
+            if record.expected_price < 0:
+                raise ValidationError("Expected price must be greater than 0")
```
We already have constrain, so do we need it here?

hapat-odoo committed:
[IMP] estate: added Inline Views, widgets, list order

-Implemented inline views and widgets to enhance the user interface within the
Real Estate module.
-Introduced list-ordering functionality to improve the organization and sorting
of listings.
-Added attribute and option configurations to increase customization capabilities
and overall flexibility for users.

bit-odoo commented on Estate/models/estate_property_type.py:
```diff
@@ -0,0 +1,36 @@
+    _check_type_name_unique_ratio = models.Constraint(
+        'CHECK(name)',
+        'The property name must be unique.'
+    )
```
Should be one empty line at the end of the file.

bit-odoo reviewed (commented):
Hello!
I have made another round of review
As we discussed early, can you please update your commit message title and please update your PR title and message?

hapat-odoo committed:
[IMP] estate:  Implemented inheritance logic

-implemented inheritance in models,views
-implemented some ui changes

bit-odoo commented on Estate/models/estate_property.py:
```diff
@@ -0,0 +1,110 @@
+    garden_area = fields.Integer()
+    garden_orientation = fields.Selection([("north", "North"),("south", "South"),("east", "East"),("west", "West"),])
+    active = fields.Boolean(default=True)
+    state = fields.Selection([("new", "New"),("offer_received", "Offer Received"),("offer_accepted", "Offer Accepted"),("sold", "Sold"),("canceled", "Canceled")], required=True, copy=False , )
```
```suggestion
    state = fields.Selection(
        [
            ("new", "New"),
            ("offer_received", "Offer Received"),
            ("offer_accepted", "Offer Accepted"),
            ("sold", "Sold"),
            ("canceled", "Canceled")
        ], required=True, copy=False)
```
For better readability.

bit-odoo commented on Estate/models/estate_property.py:
```diff
@@ -0,0 +1,110 @@
+    total_area = fields.Float(compute="_compute_total_area", string="Total Area", store=True)
+    best_price = fields.Float(compute="_compute_best_price", string="Best Offer", store=True)
+    validity_days = fields.Integer(default=7)
+    date_deadline = fields.Date(compute="_compute_date_deadline",inverse="_inverse_date_deadline",store=True,)
```
```suggestion
    date_deadline = fields.Date(compute="_compute_date_deadline", inverse="_inverse_date_deadline", store=True,)
```

bit-odoo commented on Estate/models/estate_property.py:
```diff
@@ -0,0 +1,110 @@
+    _check_price = models.Constraint(
+        'CHECK(expected_price > 0 AND selling_price >= 0)',
+        'The expected price of a property must be strictly positive.'
+    )
```
EOL missing.

bit-odoo commented on Estate/views/estate_property_views.xml:
```diff
@@ -0,0 +1,123 @@
+    <field name="arch" type="xml">
+        <form string="Property">
+            <header>
+                <button name="action_set_next_status" string="Next" type="object"/>
```
I do not think that this button is necessary. As a user can not change the state manually every time.

bit-odoo commented on Estate/models/estate_property.py:
```diff
@@ -0,0 +1,110 @@
+                raise ValidationError("You cannot delete a new or canceled property.")
+        return super().unlink()
+    
+    _check_price = models.Constraint(
```
It should be after field declaration, as I mentioned in the earlier comment.

bit-odoo commented on Estate/models/estate_property_type.py:
```diff
@@ -0,0 +1,33 @@
+        'CHECK(name)',
+        'The property name must be unique.'
+    )
+    
```
EOL missing.

bit-odoo commented on Estate/views/estate_property_offer_views.xml:
```diff
@@ -0,0 +1,34 @@
+            </form>
+        </field>
+    </record>
+</odoo>
```
EOL missing.

bit-odoo commented on Estate/views/estate_property_views.xml:
```diff
@@ -0,0 +1,123 @@
+    </record>
+
+        
+</odoo>
```
EOL missing.

bit-odoo commented on Estate/views/res_users_view.xml:
```diff
@@ -0,0 +1,25 @@
+          <field name="view_mode">form</field>
+          <field name="view_id" ref="view_users_form_inherit_properties"/>
+    </record>
+</odoo>
```
EOL missing.

bit-odoo reviewed (commented):
Hello,
I think you have to adapt the changes that i suggest you. I have to make comment multiple times it is not good.

hapat-odoo committed:
[IMP] estate: Refine model fields and view structure

- Clean whitespace and improve formatting across model files.
- Enhance state management logic in estate_property.
- Remove unused buttons and fields to simplify property views.
- Update selection fields for improved readability and maintainability.

hapat-odoo committed:
[IMP] estate:  Implemented Interact with other modules

-Enhance property management with offer handling and invoice creation
